### PR TITLE
nvuApp: pdf created from questionnaires

### PR DESCRIPTION
**Pull request by p-nedelchev**
Questionnaires can be exported as pdf files.

### CODE REVIEW

**Comment by p-nedelchev** [comment]
PR is up to date.

**Comment by p-nedelchev** [comment]
Some SS:
questionnaire
screenshot from 2017-02-06 11-43-56

File: src/main/java/com/clouway/nvuapp/adapter/http/controllers/PdfHandler.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+public class PdfHandler implements SecuredHandler {
+    private final QuestionnaireRepository repository;
+
+    public PdfHandler(QuestionnaireRepository repository) {
+        this.repository = repository;
+    }
+
+    @Override
+    public Response handle(Request request, Tutor tutor) {
+        String id = request.param("id");
```
Comment (mgenov) [inline src/main/java/com/clouway/nvuapp/adapter/http/controllers/PdfHandler.java]: Someone could pass questionary that is not available, so this case should be handled.

File: src/main/java/com/clouway/nvuapp/adapter/http/controllers/PdfHandler.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        Questionnaire questionnaire = repository.getQuestionnaire(Integer.parseInt(id));
+        PdfGenerator generator = new PdfGenerator();
```
Comment (mgenov) [inline src/main/java/com/clouway/nvuapp/adapter/http/controllers/PdfHandler.java]: I don't see why you have extracted Generator interface (to be able to mock it) but you are not using it.

File: src/main/resources/com/clouway/nvuapp/adapter/http/servlet/login.html [hunk 3]
```diff
@@ -6,11 +6,11 @@
...
 </head>
 <body>
-<form class="loginForm">
+<form class="loginForm" method="post">
     <input placeholder="ID" id="loginid" type="text" name="LoginID" size="16"/>
     <input placeholder="Password" id="loginpass" type="password" name="LoginPass" size="16"/>
-    <input type="submit" onsubmit="action='/login' method= 'post'" value="Вход"/>
-    <input type="submit" onsubmit="action='/registration' method= 'post'" value="Регистрация"/>
+    <input type="submit" formaction="/login" value="Вход"/>
+    <input type="submit" formaction="/registration" value="Регистрация"/>
```
Comment (mgenov) [inline src/main/resources/com/clouway/nvuapp/adapter/http/servlet/login.html]: Link to the registration page could be with a simple Link. 

```html
<a href="/registration">Registration</a>
```
as it doesn't require post information to be passed.

File: src/main/resources/com/clouway/nvuapp/adapter/http/servlet/showTest.html [hunk 4]
```diff
@@ -9,6 +9,9 @@
 <#include "/adminNavbar.ftl">
     <div class="questionList">
     <h1>Въпросник № ${num}</h1>
+        <form action="/pdf" target="_blank">
```
Comment (mgenov) [inline src/main/resources/com/clouway/nvuapp/adapter/http/servlet/showTest.html]: You don't need forms for this I suppose. 
```html
<a href="/pdf?questionary=${num}" target="_blank">${num}"</a>

```

File: src/main/java/com/clouway/nvuapp/core/Generator.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.clouway.nvuapp.core;
+
+import java.io.ByteArrayOutputStream;
+
+/**
+ * @author Petar Nedelchev (peter.krasimirov@gmail.com)
+ */
+public interface Generator {
+    ByteArrayOutputStream generate(Questionnaire questionnaire);
```
Comment (mgenov) [inline src/main/java/com/clouway/nvuapp/core/Generator.java]: // void generate(Questionnaire questionnaire, OutputStream output); //

byte[] generate(Questionnaire questionnaire);

File: src/main/java/com/clouway/nvuapp/core/PdfGenerator.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * @author Petar Nedelchev (peter.krasimirov@gmail.com)
+ */
+public class PdfGenerator implements Generator {
+    private final ByteArrayOutputStream baos;
+    private static final String FONT = "src/main/resources/Times New Roman.ttf";
```
Comment (mgenov) [inline src/main/java/com/clouway/nvuapp/core/PdfGenerator.java]: Use classloader.

File: src/main/java/com/clouway/nvuapp/adapter/http/controllers/PdfHandler.java (hunk 1 above)
Comment (p-nedelchev) [inline src/main/java/com/clouway/nvuapp/adapter/http/controllers/PdfHandler.java]: I handled cases where questionnaire with given id doesn't exist or invalid id is passed.

File: src/main/java/com/clouway/nvuapp/adapter/http/controllers/PdfHandler.java (hunk 2 above)
Comment (p-nedelchev) [inline src/main/java/com/clouway/nvuapp/adapter/http/controllers/PdfHandler.java]: I use the interface now.

File: src/main/resources/com/clouway/nvuapp/adapter/http/servlet/login.html (hunk 3 above)
Comment (p-nedelchev) [inline src/main/resources/com/clouway/nvuapp/adapter/http/servlet/login.html]: Change form to simple link.

File: src/main/resources/com/clouway/nvuapp/adapter/http/servlet/showTest.html (hunk 4 above)
Comment (p-nedelchev) [inline src/main/resources/com/clouway/nvuapp/adapter/http/servlet/showTest.html]: Changed to simple link.

File: src/main/java/com/clouway/nvuapp/core/Generator.java (hunk 5 above)
Comment (p-nedelchev) [inline src/main/java/com/clouway/nvuapp/core/Generator.java]: Method signature is like this  `byte [] generate(Questionnaire questionnaire);` now

File: src/main/java/com/clouway/nvuapp/core/PdfGenerator.java (hunk 6 above)
Comment (p-nedelchev) [inline src/main/java/com/clouway/nvuapp/core/PdfGenerator.java]: Used `PdfGenerator.class.getResource("TimesNewRoman.ttf").getFile()`

**Comment by p-nedelchev** [comment]
PR is up to date.

File: src/main/java/com/clouway/nvuapp/adapter/http/controllers/PdfHandler.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+    public PdfHandler(QuestionnaireRepository repository, Generator pdfGenerator) {
+        this.repository = repository;
+        this.pdfGenerator = pdfGenerator;
+    }
+
+    @Override
+    public Response handle(Request request, Tutor tutor) {
+        String id = request.param("id");
+        if(! isNumeric(id)) {
```
Comment (mgenov) [inline src/main/java/com/clouway/nvuapp/adapter/http/controllers/PdfHandler.java]: formatting required.

File: src/main/java/com/clouway/nvuapp/core/PdfGenerator.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        }
+        document.add(answersList);
+        document.close();
+        return baos.toByteArray();
+    }
+
+    private PdfFont getTimesNewRomanFont() {
+        PdfFont font = null;
+        String timesNewRoman = PdfGenerator.class.getResource("TimesNewRoman.ttf").getFile();
+        System.out.println(timesNewRoman);
```
Comment (VasilMitovClouway) [inline src/main/java/com/clouway/nvuapp/core/PdfGenerator.java]: Remove sout from production code.

File: src/main/resources/com/clouway/nvuapp/adapter/http/servlet/notAvailableQuestionnaire.html [hunk 9]
```diff
@@ -0,0 +1,12 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <title>Грешка</title>
+</head>
+<body>
+    <div>
```
Comment (VasilMitovClouway) [inline src/main/resources/com/clouway/nvuapp/adapter/http/servlet/notAvailableQuestionnaire.html]: We could add the adminNavbar.ftl here, or some other form of navigation.

File: src/main/java/com/clouway/nvuapp/adapter/http/controllers/PdfHandler.java (hunk 7 above)
Comment (p-nedelchev) [inline src/main/java/com/clouway/nvuapp/adapter/http/controllers/PdfHandler.java]: Formatted.

File: src/main/java/com/clouway/nvuapp/core/Generator.java [hunk 10]
```diff
@@ -0,0 +1,8 @@
+package com.clouway.nvuapp.core;
+
+/**
+ * @author Petar Nedelchev (peter.krasimirov@gmail.com)
+ */
+public interface Generator {
```
Comment (p-nedelchev) [inline src/main/java/com/clouway/nvuapp/core/Generator.java]: Method renamed.

File: src/main/java/com/clouway/nvuapp/core/PdfGenerator.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+import com.itextpdf.layout.property.ListNumberingType;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * @author Petar Nedelchev (peter.krasimirov@gmail.com)
+ */
+public class PdfGenerator implements Generator {
```
Comment (p-nedelchev) [inline src/main/java/com/clouway/nvuapp/core/PdfGenerator.java]: Renamed to suggested name.

File: src/main/java/com/clouway/nvuapp/core/PdfGenerator.java (hunk 8 above)
Comment (p-nedelchev) [inline src/main/java/com/clouway/nvuapp/core/PdfGenerator.java]: Removed sout from code.

File: src/test/java/com/clouway/nvuapp/adapter/http/controllers/PdfHandlerTest.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+ */
+public class PdfHandlerTest {
+    @Rule
+    public JUnitRuleMockery context = new JUnitRuleMockery();
+
+    @Test
+    public void happyPath() throws Exception {
+        Generator pdfGenerator = context.mock(Generator.class);
+
+        Questionnaire questionnaire = new Questionnaire(1);
```
Comment (p-nedelchev) [inline src/test/java/com/clouway/nvuapp/adapter/http/controllers/PdfHandlerTest.java]: Test simplified.

File: src/main/resources/com/clouway/nvuapp/adapter/http/servlet/notAvailableQuestionnaire.html (hunk 9 above)
Comment (p-nedelchev) [inline src/main/resources/com/clouway/nvuapp/adapter/http/servlet/notAvailableQuestionnaire.html]: adminNavbar.ftl was added.

**Comment by p-nedelchev** [comment]
PR is up to date.

**Comment by p-nedelchev** [comment]
Reminder. Needs another review.

File: src/test/java/com/clouway/nvuapp/adapter/http/controllers/PdfHandlerTest.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+ */
+public class PdfHandlerTest {
+    @Rule
+    public JUnitRuleMockery context = new JUnitRuleMockery();
+
+    @Test
+    public void happyPath() throws Exception {
+        QuestionnairePdfGenerator pdfGenerator = context.mock(QuestionnairePdfGenerator.class);
+        QuestionnaireRepository repository = context.mock(QuestionnaireRepository.class);
+        Request request = context.mock(Request.class);
```
Comment (mgenov) [inline src/test/java/com/clouway/nvuapp/adapter/http/controllers/PdfHandlerTest.java]: I think that you already have MockRequest about this, so you don't have to add another jmock expectation.

Comment (p-nedelchev) [inline src/test/java/com/clouway/nvuapp/adapter/http/controllers/PdfHandlerTest.java]: Yes sorry I used FakeRequest now.

**Comment by p-nedelchev** [comment]
PR is up to date.

**Pull request merged**